lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -727,33 +728,68 @@ namespace Js
     {
         EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
 
+#if defined(INTL_ICU)
+        INTL_CHECK_ARGS(args.Info.Count == 2 && JavascriptString::Is(args[1]));
+
+        UErrorCode status = U_ZERO_ERROR;
+        JavascriptString *langtag = JavascriptString::UnsafeFromVar(args[1]);
+        utf8::WideToNarrow langtag8(langtag->GetSz(), langtag->GetLength());
+
+        // ICU doesn't have a full-fledged canonicalization implementation that correctly replaces all preferred values
+        // and grandfathered tags, as required by #sec-canonicalizelanguagetag.
+        // However, passing the locale through uloc_forLanguageTag -> uloc_toLanguageTag gets us most of the way there
+        // by replacing some(?) values, correctly capitalizing the tag, and re-ordering extensions
+        int parsedLength = 0;
+        char localeID[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int forLangTagResultLength = uloc_forLanguageTag(langtag8, localeID, ULOC_FULLNAME_CAPACITY, &parsedLength, &status);
+        if (status == U_ILLEGAL_ARGUMENT_ERROR)
+        {
+            // The string passed in to NormalizeLanguageTag has already passed IsStructurallyValidLanguageTag.
+            // However, duplicate unicode extension keys, such as "de-u-co-phonebk-co-phonebk", are structurally
+            // valid according to RFC5646 yet still trigger U_ILLEGAL_ARGUMENT_ERROR
+            // V8 ~6.2 says that the above language tag is invalid, while SpiderMonkey ~58 handles it.
+            // Until we have a more spec-compliant implementation of CanonicalizeLanguageTag, err on the side
+            // of caution and say it is invalid.
+            // NOTE: make sure we check for `undefined` at the platform.normalizeLanguageTag callsite.
+            return scriptContext->GetLibrary()->GetUndefined();
+        }
+
+        // forLangTagResultLength can be 0 if langtag is "und".
+        // uloc_toLanguageTag("") returns "und", so this works out (forLanguageTag can return >= 0 but toLanguageTag must return > 0)
+        ICU_ASSERT(status, forLangTagResultLength >= 0 && ((charcount_t) parsedLength) == langtag->GetLength());
+
+        char canonicalized[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int toLangTagResultLength = uloc_toLanguageTag(localeID, canonicalized, ULOC_FULLNAME_CAPACITY, true, &status);
+        ICU_ASSERT(status, toLangTagResultLength > 0);
+
+        // allocate toLangTagResultLength + 1 to leave room for null terminator
+        char16 *canonicalized16 = RecyclerNewArrayLeaf(scriptContext->GetRecycler(), char16, toLangTagResultLength + 1);
+        size_t canonicalized16Len = 0;
+        HRESULT hr = utf8::NarrowStringToWideNoAlloc(
+            canonicalized,
+            toLangTagResultLength,
+            canonicalized16,
+            toLangTagResultLength + 1,
```

Comment:
>toLangTagResultLength + 1 [](start = 12, length = 25)

since this expression is used twice and in both cases is semantically the length of canonicalized16, can you make this a local?

Reply:
This is the same problem as elsewhere in the file where we sometimes have to deal with the null terminator and sometimes don't. I can make a variable for it though.